python/tritonserver/_api/_datautils.py

```diff
@@ -0,0 +1,327 @@
+    lambda: None,
+    {
+        _dlpack.DLDeviceType.kDLCUDA: triton_bindings.TRITONSERVER_MemoryType.GPU,
+        _dlpack.DLDeviceType.kDLCPU: triton_bindings.TRITONSERVER_MemoryType.CPU,
+    },
+)
+
+DLPACK_TO_TRITON_DTYPE = defaultdict(
+    lambda: triton_bindings.TRITONSERVER_DataType.INVALID,
+    {
+        (_dlpack.DLDataTypeCode.kDLBool, 1): triton_bindings.TRITONSERVER_DataType.BOOL,
+        (
+            _dlpack.DLDataTypeCode.kDLFloat,
+            16,
+        ): triton_bindings.TRITONSERVER_DataType.FP16,
+    },
+)
+
+
+NUMPY_TO_TRITON_DTYPE = defaultdict(
+    lambda: triton_bindings.TRITONSERVER_DataType.INVALID,
+    {
+        bool: triton_bindings.TRITONSERVER_DataType.BOOL,
+        numpy.int8: triton_bindings.triton_bindings.TRITONSERVER_DataType.INT8,
+        numpy.int16: triton_bindings.TRITONSERVER_DataType.INT16,
+        numpy.int32: triton_bindings.TRITONSERVER_DataType.INT32,
+        numpy.int64: triton_bindings.TRITONSERVER_DataType.INT64,
+        numpy.uint8: triton_bindings.TRITONSERVER_DataType.UINT8,
+        numpy.uint16: triton_bindings.TRITONSERVER_DataType.UINT16,
+        numpy.uint32: triton_bindings.TRITONSERVER_DataType.UINT32,
+        numpy.uint64: triton_bindings.TRITONSERVER_DataType.UINT64,
+        numpy.float16: triton_bindings.TRITONSERVER_DataType.FP16,
+        numpy.float32: triton_bindings.TRITONSERVER_DataType.FP32,
+        numpy.float64: triton_bindings.TRITONSERVER_DataType.FP64,
+        numpy.bytes_: triton_bindings.TRITONSERVER_DataType.BYTES,
+        numpy.str_: triton_bindings.TRITONSERVER_DataType.BYTES,
+        numpy.object_: triton_bindings.TRITONSERVER_DataType.BYTES,
+    },
+)
+
+TRITON_TO_NUMPY_DTYPE = defaultdict(
+    lambda: triton_bindings.TRITONSERVER_DataType.INVALID,
+    {value: key for key, value in NUMPY_TO_TRITON_DTYPE.items()},
+)
+
+
+class MemoryAllocator(ABC):
+    @abstractmethod
+    def allocate(
+        self,
+        allocator,
+        tensor_name,
+        byte_size,
+        memory_type,
+        memory_type_id,
+        user_object,
+    ):
+        pass
+
+    @abstractmethod
+    def release(
+        self,
+        allocator,
+        buffer_,
+        buffer_user_object,
+        byte_size,
+        memory_type,
+        memory_type_id,
+    ):
+        pass
+
+    @abstractmethod
+    def start(self, allocator, user_object):
+        pass
+
+    def create_response_allocator(self):
+        self._allocator = triton_bindings.TRITONSERVER_ResponseAllocator(
+            self.allocate, self.release, self.start
+        )
+
+        if hasattr(self, "query_preferred_memory_type"):
+            self._allocator.set_query_function(self.query_preferred_memory_type)
+
+        if hasattr(self, "set_attributes"):
+            self._allocator.set_buffer_attributes_function(self.set_buffer_attributes)
+        return self._allocator
+
+
+try:
+    import cupy
+
+    class CupyAllocator(MemoryAllocator):
```

Comment:
question: where is this Allocator class used?

Reply:
Used when the request is created and callbacks are set. Currently always set to default Numpy - but plan to add an option to choose Cupy. 

Question: currently there is no negotiation - meaning the client sets the memory type - would it be better to try to meet the backends request?

Reply:
Maybe better to move the "reference" allocators to dedicated files, user can choose from "allocators.CupyAllocator" etc.

RE negotiation: it would be cool to show a collective allocator where it contains an instance of CupyAllocator and DefaultAllocator, and it delegates the allocation to them based on backend's request.

Reply:
The defaultallocator actually has both Numpy and Cupy and delegates based on the backend request (if cupy is available). One question: is query required for negotiation or is to respond with the allocation type as requested sufficient?